docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,202 @@
+Support additions to the resource list by applying the new added resources to both new and existing matching clusters
+Support both json and yaml resources
+
+### Non-Goals/Future Work
+
+Replace or compete with the Cluster Addons subproject
+Support deletion of resources. Deleting resources from the list will not result in deletion of those resources from the synced clusters, but deleted resources will not be applied to the new clusters
+Responsible for cleaning up resources applied by ClusterResourceSet controller when ClusterResourceSet resource is deleted
+Responsible for lifecycle management of the installed resources (such as CNI)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce" / "Sync"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+```
+
+Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. "Sync" mode will include reapplying the resources on resource hash change or periodically following an interval.
```

Comment:
```suggestion
Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. In the future, we may consider adding a "Sync" mode that reapplies the resources on resource hash change and/or periodically.
```